### ISSUE TITLE
`type[Any]` cannot be passed to `Optional[]`

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2687]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

import inspect
from typing import Any, Optional, cast


def foo(bar: int | str = None) -> None:
    ...


def resolve_annotation(
    annotation: Any, globals: dict[str, Any]
) -> type[Any]:
    '''Similar to `typing._eval_type()` but handles `X|Y` syntax in 3.9'''
    ...


signature = inspect.signature(foo)
bar = signature.parameters['bar']

annotation = resolve_annotation(bar.annotation, cast(Any, foo).__globals__)

if annotation is not inspect._empty and bar.default is None:
    optional_annotation = Optional[annotation]
```

Two errors are reported for `annotation`:

1. Illegal type annotation: variable not allowed unless it is a type alias
2. Expected class type but received "type"

One error is reported for `optional_annotation`:

1. Type of "optional_annotation" is partially unknown
  Type of "optional_annotation" is "Unknown | None"

**Expected behavior**
No errors should be reported. In addition, I wonder if the type of `optional_annotation` could be inferred as `type[Optional[type[Any]]]`.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.194

**Comment by erictraut** [comment]
I think this behavior is correct. Variables cannot be used as type arguments. Mypy issues the same error.

**Comment by bryanforbes** [comment]
Should I `# type: ignore` the line? (I really hate doing that)

**Comment by erictraut** [comment]
I'm not sure what you're trying to do. It looks like you're expecting that `Optional` to be an actual class type rather than a special form, as it's defined in typing.pyi and in PEP 484.

**Comment by bryanforbes** [comment]
In the library I'm writing, I'm using type annotations at runtime to determine some stuff (I know, super technical). Normally, I'd be able to use `typing.get_type_hints()`, but 3.10 added the `X | Y` syntax which can be used in 3.9 with `from __future__ import annotations`, but `typing.get_type_hints()` will throw an exception in 3.9. To work around this issue, I wrote something equivalent to `get_type_hints()` that parses the `X | Y` syntax in 3.9 (using `future_typing`). As you can see from my code example, I'm using `inspect` to get the parameter information, but the annotation that `inspect` returns is different than the annotation that `get_type_hints()` would return for parameters that have an annotation and a default value of `None`:

```python
import inspect
import typing


def foo(bar: int = None) -> None:
    ...

inspect.signature(foo).parameters['bar'].annotation  # `int` builtin
typing.get_type_hints(foo)['bar']  # `typing._UnionGenericAlias` with `__args__` set to `(int, NoneType)`
```

So in order to emulate what `typing.get_type_hints()` is doing, I looked at [the source](https://github.com/python/cpython/blob/ca08655b808aed2e3abeb64cb67d98a79a661dda/Lib/typing.py#L1847-L1848) and added something similar to my code. I know that `Optional[int]` at runtime produces an instance of `typing._UnionGenericAlias` (so my suggestion for what type a runtime `Optional[annotation]` should produce was wrong), but it doesn't seem that either mypy or pyright support type checking what I'm trying to do (which I'm fine with as an answer).

**Comment by erictraut** [comment]
Thanks for the details. Yeah, I think pyright and mypy are correct to flag a use of `typing.Optional` as an error when it's used in a dynamic expression.
